apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php

```diff
@@ -232,32 +234,45 @@ function load_wpcom_block_editor_nux() {
 add_action( 'plugins_loaded', __NAMESPACE__ . '\load_wpcom_block_editor_nux' );
 
 /**
- * Load editing toolkit block patterns from the API
+ * Return a function that loads and register block patterns from the API. This
+ * function can be registered to the `rest_dispatch_request` filter.
  *
- * @param obj $current_screen The current screen object.
+ * @param Function $register_patterns_func A function that when called will
+ * register the relevant block patterns in the registry.
  */
-function load_block_patterns_from_api( $current_screen ) {
-	if ( ! apply_filters( 'a8c_enable_block_patterns_api', false ) ) {
-		return;
-	}
-
-	$is_site_editor = ( function_exists( 'gutenberg_is_edit_site_page' ) && gutenberg_is_edit_site_page( $current_screen->id ) );
-
-	if ( ! $current_screen->is_block_editor && ! $is_site_editor ) {
-		return;
-	}
+function load_block_patterns_from_api( $register_patterns_func ) {
+	/**
+	 * Load editing toolkit block patterns from the API.
+	 *
+	 * It will only register the patterns for certain allowed requests and
+	 * return early otherwise.
+	 *
+	 * @param mixed           $response
+	 * @param WP_REST_Request $request
+	 */
+	return function ( $response, $request ) use ( $register_patterns_func ) {
+		$route           = $request->get_route();
+		$request_allowed = preg_match( '/^\/wp\/v2\/sites\/[0-9]+\/block\-patterns\/(patterns|categories)$/', $route );
 
-	$editor_type = 'block_editor';
+		if ( ! $request_allowed || ! apply_filters( 'a8c_enable_block_patterns_api', false ) ) {
```

Comment:
One interesting thing to note is that we seem to have two filters:
* `a8c_enable_block_patterns_api`
* and `a8c_enable_fse_block_patterns_api`

I didn't change any of them, they stayed as-is. At the time I write this, I don't know if the fact we have these two is intentional or if it was an oversight.

Reply:
They seem to serve different purposes, it's not oversight. `a8c_enable_block_patterns_api` was used to roll out the block patterns feature to certain percent of users, and has been at full 100% since October 2020. This filter can probably be fully removed at this point, except maybe it's disabled for unit tests. 

`a8c_enable_fse_block_patterns_api` was use for rolling out FSE, enabling it only for sites with certain stickers/options. Today it expands to just `is_fse_theme()`, i.e., FSE patterns are loaded only when the site has a block theme enabled.